giskard/utils/worker_pool.py

```diff
@@ -0,0 +1,337 @@
+        if timeout is not None:
+            self._with_timeout_tasks.append((task.id, time.monotonic() + timeout))
+        return res
+
+    def shutdown(self, wait=True, timeout: float = 5):
+        if self._state in FINAL_STATES:
+            return
+        # Changing state, so that thread will stop
+        # Killer thread will also do cleanup
+        self._state = PoolState.STOPPING
+        # Cancelling all futures we have
+        for future in self._futures_mapping.values():
+            if future.cancel() and not future.done():
+                future.set_exception(CancelledError("Executor is stopping"))
+        # Emptying running_tasks queue
+        while not self._running_tasks_queue.empty():
+            try:
+                self._running_tasks_queue.get_nowait()
+            except BaseException:
+                pass
+        # Try to nicely stop the worker, by adding None into the running tasks
+        for _ in range(self._nb_workers):
+            self._running_tasks_queue.put(None, timeout=1)
+        # Wait for process to stop by themselves
+        p_list = list(self._processes.values())
+        if wait:
+            _wait_process_stop(p_list, timeout=timeout)
+
+        # Clean all the queues
+        for queue in [self._pending_tasks_queue, self._tasks_results, self._running_tasks_queue]:
+            # In python 3.8, Simple queue seems to not have close method
+            if hasattr(queue, "close"):
+                queue.close()
+        # Waiting for threads to finish
+        for t in self._threads:
+            t.join(timeout=1)
+            if t.is_alive():
+                LOGGER.warning("Thread %s still alive at shutdown", t.name)
+        # Changing state to stopped
+        self._state = PoolState.STOPPED
+        # Cleaning up processes
+        exit_codes = _stop_processes(p_list, timeout=1)
+        self._manager.shutdown()
+        return exit_codes
+
+
+def _results_thread(
+    executor: WorkerPoolExecutor,
+):
+    # Goal of this thread is to feed the running tasks from pending one as soon as possible
+    while True:
+        while executor._state not in FINAL_STATES and (executor._tasks_results.empty()):
+            # TODO(Bazire): find a way to improve this ?
+            # Cannot use select, since we want to be windows compatible
+            sleep(0.01)
+        if executor._state in FINAL_STATES:
+            return
+        result = executor._tasks_results.get()
+        future = executor._futures_mapping.get(result.id)
+        if future.cancelled():
+            try:
+                del executor._futures_mapping[result.id]
+            except BaseException:
+                pass
+            continue
+        future.logs = result.logs
+        if result.exception is None:
+            future.set_result(result.result)
+        else:
+            # TODO(Bazire): improve to get Traceback
+            future.set_exception(RuntimeError(result.exception))
```

Comment:
why can't it by directly `future.set_exception(result.exception)` ?

Reply:
Exception is a string in this case (the exception + traceback), because it needs to be pickable, and exception are linked to many objects + have circular references